Modules/posixmodule.c

```diff
@@ -4800,43 +4802,70 @@ os__path_isdir_impl(PyObject *module, PyObject *path)
     }
 
     Py_BEGIN_ALLOW_THREADS
-    if (_path.fd != -1) {
-        hfile = _Py_get_osfhandle_noraise(_path.fd);
-        close_file = FALSE;
-    }
-    else {
-        hfile = CreateFileW(_path.wide, FILE_READ_ATTRIBUTES, 0, NULL,
-                            OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
+    if (_path.wide) {    
+        if (_Py_GetFileInformationByName(_path.wide, FileStatBasicByNameInfo,
+                                         &statInfo, sizeof(statInfo))) {
+            if (!(statInfo.FileAttributes & FILE_ATTRIBUTE_REPARSE_POINT)) {
+                slow_path = FALSE;
+                result = statInfo.FileAttributes & FILE_ATTRIBUTE_DIRECTORY;
+            } else if (!(statInfo.FileAttributes & FILE_ATTRIBUTE_DIRECTORY)) {
+                slow_path = FALSE;
+                result = 0;
+            }
+        } else {
+            switch(GetLastError()) {
+                case ERROR_FILE_NOT_FOUND:
+                case ERROR_PATH_NOT_FOUND:
+                case ERROR_NOT_READY:
+                case ERROR_BAD_NET_NAME:
+                    /* These errors aren't worth retrying with the slow path */
+                    slow_path = FALSE;
+                    result = 0;
+                case ERROR_NOT_SUPPORTED:
+                    /* indicates the API couldn't be loaded */
+                    break;
+            }
+        }
     }
-    if (hfile != INVALID_HANDLE_VALUE) {
-        if (GetFileInformationByHandleEx(hfile, FileBasicInfo, &info,
-                                         sizeof(info)))
-        {
-            result = info.FileAttributes & FILE_ATTRIBUTE_DIRECTORY;
+    if (slow_path) {
+        if (_path.fd != -1) {
+            hfile = _Py_get_osfhandle_noraise(_path.fd);
+            close_file = FALSE;
```

Comment:
A couple of bugs slipped by when the new accelerator functions were added. These bugs should probably be resolved in a separate issue before this pull request gets merged.

* If `_Py_get_osfhandle_noraise()` fails (e.g. a bad file descriptor), the result is `INVALID_HANDLE_VALUE`, but the thread's last error isn't set and is thus a random error code. It could be one of the errors that's handled by calling `STAT()`, but in this case there is no `_path.wide` value to check. It happens that `CreateFileW()` doesn't raise an OS exception that crashes the process when passed a null pointer for `lpFileName`, but that's not documented and shouldn't be relied on. 
* If `GetFileType(hfile)` isn't `FILE_TYPE_DISK`, we have to return a false result. Since we didn't open the file, we don't know whether or not it has a pending synchronous operation that's blocked indefinitely. For example, a pipe or character file could have a pending synchronous read that may never complete. In this case, `GetFileInformationByHandleEx()` would block.